### PR TITLE
Slashed validators were applied

### LINKED ISSUE #487: preimages table default key constraint violation
when re-enrollment is done at height 20. then preimages at height 21  violation the primary key constraint

.

### PR DESCRIPTION
**Pull request by TrustHenry**
If there is no pre-image in the block header, it is slashed.

### CODE REVIEW

File: src/modules/storage/LedgerStorage.ts [hunk 1]
```diff
@@ -907,20 +907,54 @@ export class LedgerStorage extends Storages {
...
         }
 
+        function updateValidatorSlashing(storage: LedgerStorage, height: Height, utxo_key: Hash, enrolled_at: Height) {
+            return new Promise<void>((resolve, reject) => {
+                storage
+                    .query(
+                        `UPDATE validators
+                            SET slashed = ? ,
+                            slashed_height = ?
+                            WHERE utxo_key = ? AND enrolled_at = ?`,
```
Comment (MichaelKim20) [inline src/modules/storage/LedgerStorage.ts:941]: It changed from `address` to `utxo_key`.
Good

File: src/modules/storage/LedgerStorage.ts [hunk 2]
```diff
@@ -907,20 +907,54 @@ export class LedgerStorage extends Storages {
...
         return new Promise<void>((resolve, reject) => {
             (async () => {
                 const validators: any[] = await this.getValidatorsAPI(block.header.height, null, conn);
                 for (let idx = 0; idx < block.header.preimages.length; idx++) {
+                    if (idx >= validators.length) continue;
+                    if (validators[idx] === undefined) continue;
                     try {
-                        if (undefined !== validators[idx] && block.header.preimages[idx] !== new Hash(Buffer.alloc(Hash.Width))) {
-                            await save_preimage(
+                        if (Buffer.compare(block.header.preimages[idx].data, Buffer.alloc(Hash.Width)) === 0) {
```
Comment (MichaelKim20) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
                        if (Hash.isNull(block.header.preimages[idx])) {
```
Please raise the version of SDK to 0.2.6, and I hope you can modify the code like this.

File: src/modules/storage/LedgerStorage.ts [hunk 3]
```diff
@@ -2888,75 +2898,25 @@ export class LedgerStorage extends Storages {
...
-                slashedValidators = cycleValidators.filter(v => !block_preimages.some((v1) => v.address === v1.address));
-
-                for (let i = 0; i < slashedValidators.length; i++) {
-                    await updateValidatorSlashing(this, block.header.height, slashedValidators[i].address, slashedValidators[i].enrolled_at);
-                }
-
-                slashedValidators.forEach((e) => {
-                    cycleValidators.splice(cycleValidators.indexOf(e), 1);
-                });
+                const cycleValidators: any[] = await this.getValidatorsAPI(block.header.height, null);
```
Comment (MichaelKim20) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
                const cycleValidators: any[] = await this.getValidatorsAPI(block.header.height, null, conn);
```

**Review by MichaelKim20 — CHANGES_REQUESTED** [review]
I've reviewed it so far.
Let me check more.

File: src/modules/storage/LedgerStorage.ts (hunk 3 above)
Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: I applied the suggestion.

File: src/modules/storage/LedgerStorage.ts (hunk 2 above)
Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: I applied the suggestion.

Comment (MichaelKim20) [reply src/modules/storage/LedgerStorage.ts]: The suggestion above is wrong.
I think it'd be good to do it like this.
```suggestion
                        if (block.header.preimages[idx].isNull()) {
```

Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: I applied the suggestion.

**Review by MichaelKim20 — APPROVED** [review]
LGTM

**Pull request merged**